File: coffeeshop-frontend/src/app/services/event.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import {
  EventResponseDto,
  EventCreateRequest,
  EventUpdateRequest,
  PageResponseDto,
  EventSearchParams,
} from '../models/event.model';

@Injectable({ providedIn: 'root' })
export class EventService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/api/v1/event`;

  getAll(): Observable<EventResponseDto[]> {
    return this.http.get<EventResponseDto[]>(this.base);
  }

  search(params: EventSearchParams = {}): Observable<PageResponseDto<EventResponseDto>> {
    const httpParams: Record<string, string> = {
      page: String(params.page ?? 0),
      size: String(params.size ?? 10),
    };
    if (params.q?.trim()) {
      httpParams['q'] = params.q.trim();
    }
    if (params.dateFrom?.trim()) {
      httpParams['dateFrom'] = params.dateFrom.trim();
    }
    if (params.dateTo?.trim()) {
      httpParams['dateTo'] = params.dateTo.trim();
    }
    return this.http.get<PageResponseDto<EventResponseDto>>(this.base, { params: httpParams });
  }

  getByShopId(shopId: string): Observable<EventResponseDto[]> {
    return this.http.get<EventResponseDto[]>(this.base, { params: { shopId } });
  }

  getById(eventId: string): Observable<EventResponseDto> {
    return this.http.get<EventResponseDto>(`${this.base}/${eventId}`);
  }

  create(req: EventCreateRequest): Observable<EventResponseDto> {
    return this.http.post<EventResponseDto>(this.base, req);
  }

  update(eventId: string, req: EventUpdateRequest): Observable<EventResponseDto> {
    return this.http.put<EventResponseDto>(`${this.base}/${eventId}`, req);
  }

  delete(eventId: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/${eventId}`);
  }
}

File: coffeeshop/src/main/java/com/coffeeshop/coffeeshop/service/EventService.java
package com.coffeeshop.coffeeshop.service;

import com.coffeeshop.coffeeshop.model.Event;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface EventService {

    List<Event> findAll();

    Page<Event> search(Optional<String> query, Optional<LocalDate> dateFrom, Optional<LocalDate> dateTo, Pageable pageable);

    List<Event> findByShopId(UUID shopId);

    Event getById(String eventId);

    Event create(Event entity);

    Event update(String eventId, Event entity);

    void deleteById(String eventId);
}
